text/0000-orderer-pruning-and-snapshot.md

```diff
@@ -0,0 +1,275 @@
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+## Alternatives considered
+
+- Implement only pruning without the snapshot functionality.
+  This still requires storing some information about the history of the ledger, as a bare minimum we need the last config block and the hash of the last pruned block. So why should we invent a new mechanism to support pruning, we can as well use the snapshot format to store the required info.
+- Store the information about the first known block in a new field (firstBlockNum) somewhere in the blockStore or fileLedger structure.
+  Again, we already have a code implemented that stores this info in BootstrappingSnapshotInfo structure. It is better to rely on the already tested code to achieve the same functionality.
+
+# Rationale
+
+Strictly speaking to achieve the main goal - orderer ledger pruning - we don't need the snapshot functionality. But due to the implementation reasons the notion of snapshot arises in some or another way. For example we need to store somewhere the last config block and the number of the last pruned block (or the number of the first block existing in the pruned ledger). So we decided that it is better to rely on the already existing code supporting snapshots and slightly modify it to suit the needs of the OSN snapshot.
+
+In addition to the above, joining a channel from a snapshot solves many problems with the existing join-block implementation from RFC-0004 "Channel participation API without system channel", namely:
+- the need to pull potentially huge number of the historical blocks
+- the fact that the config in the join-block may be outdated and not representing the latest config
+- the need to keep the notion of "join-block" which is not present in peer logic. On the contrary, snapshot logic can serve the same purpose as "join-block" logic and at the same time unify some parts of the orderer and peer code.
+
+So we see snapshot logic as being superior to the "join-block" logic, hence in this RFC we propose to fully replace it and deprecate existing channel participation API in favor of new "join from snapshot" functionality.
+
+Incremental snapshot capability is not implemented for two main reasons:
+1. Current snapshot size is small enough that the incremental feature is not required,
+2. The design of the peer ledger snapshot does not include the incremental feature and we want to be compatible with the peer snapshot functionality at least on the conceptual level.
+
+# Prior art
+[prior-art]: #prior-art
+
+RFC-0000 "Ledger Checkpointing"
+
+RFC-0004 "Channel participation API without system channel"
+
+# Testing
+[testing]: #testing
+
+Integration tests will cover the following scenarios:
```

Comment:
You may add one more test to the list -- an orderer can produce a snapshot after itself being bootstrapped from a snapshot irrespective of whether a config block has been added afterward.

Reply:
Done, new test case has been added

Reply:
One more point that I missed earlier - you may like to add a line about the role that would be authorized to perform this command and add a test for that.

Reply:
Admin role explicit mention - done, here:
f6e34c0b072d3182a1adddc1e965d08b67310abe

As for the test - the test is to check that with the wrong identity we fail to create a snapshot. Currently Admin identity is implemented as TLS credentials, so the test is basically that with wrong TLS credentials the API call will fail - sounds like a unit test to me, not an integration test, hence we do not mention this test it explicitly in the RFC.